packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -411,32 +411,68 @@ class TextPainter {
     return _isUtf16Surrogate(prevCodeUnit) ? offset - 2 : offset - 1;
   }
 
+  // The maximum grapheme cluster length that will be attempted to have
+  // a box drawn around it before giving up. We cap this to ensure
+  // efficiency, however, very long grapheme clusters (they can be
+  // arbitrarily long) may be overlooked.
+  static const int _maxGraphemeClusterLength = 1 << 4;
```

Comment:
why do we need to have a cap? is there a degenerate case where we'd keep looking for a character even though it's not there?

Reply:
My concern was with performance, as GetRectsForRange is not a trivially fast call, and too many iterations of larger and larger sizes (should be O(N) where N=number of runs, indirectly O(#codepoints) with very large strings) could lead to skipped frames. Some text like the license files are laid out in single massive strings, and the overhead of looking though all of it NLog(N) times would likely be about the same if not more than the actual layout overhead. I imagine this could be increased to something like 1 << 7 and we will likely see no real-world differences compared to infinity. I should also cap it at the size of the actual text.

Reply:
I don't understand. Unless you really have a really long sequence of characters forming a single glyph, in which case you _want_ to go as long as the character, what's the risk? For example, the license file can never look more than 1 character regardless of what this is, no? Since the license file is all ASCII.

Reply:
I guess the license file may not be the best example. In a case such as `"blah\n blah\n .... blah\n (x30000) .... blah\n👩‍<zwj>👩‍"` with the caret position on the first 👩‍ and the affinity is upstream, this would waste a significant amount of time looking through thousands of runs of text upstream. Now this may not necessarily be slow enough for this to be a significant problem, because at that point, the layout itself would take a significant amount of time. I can remove the cap for now, and keep in mind to look out for any potential reports of it being too slow, which are unlikely to surface.

You are definitely right in that the cap is unnecessary in almost every practical situation.